File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/find-all-duplicates/code/python/find_all_duplicates_sol2.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        # Optimal: O(1) space — Negate at index to mark visited
        result = []
        for num in nums:
            idx = abs(num) - 1
            if nums[idx] < 0: result.append(idx + 1)
            else: nums[idx] = -nums[idx]
        return result

File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/find-all-duplicates/code/cpp/find_all_duplicates_sol2.cpp
class Solution {
public:
    // Optimal: O(n) time, O(1) space — Negate at index to mark visited
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> result;
        for (int i = 0; i < nums.size(); i++) {
            int idx = abs(nums[i]) - 1;     // Map value to 0-indexed position
            if (nums[idx] < 0)              // Already visited → duplicate
                result.push_back(idx + 1);
            else
                nums[idx] = -nums[idx];     // Mark visited by negating
        }
        return result;
    }
};